Cython/Compiler/ModuleNode.py

```diff
@@ -2542,117 +2527,85 @@ def generate_import_star(self, env, code):
+        code.putln('#define __Pyx_ModuleState(m) ((%s *)__Pyx_PyModule_GetState(m))' %
+                   Naming.modulestate_type_cname)
+        code.putln('#define __Pyx_Globals() __Pyx_ModuleState(%s)' % env.module_cname)
+        code.putln('#endif')
+
+    def generate_module_state_defines_begin(self, env, code):
         code.putln("#if CYTHON_USE_MODULE_STATE")
-        code.putln('#define %s %s->%s' % (
-            Naming.builtins_cname,
-            Naming.modulestateglobal_cname,
-            Naming.builtins_cname))
-        code.putln('#define %s %s->%s' % (
-            Naming.cython_runtime_cname,
-            Naming.modulestateglobal_cname,
-            Naming.cython_runtime_cname))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_tuple,
-            Naming.modulestateglobal_cname,
-            Naming.empty_tuple))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_bytes,
-            Naming.modulestateglobal_cname,
-            Naming.empty_bytes))
-        code.putln('#define %s %s->%s' % (
-            Naming.empty_unicode,
-            Naming.modulestateglobal_cname,
-            Naming.empty_unicode))
-        if Options.pre_import is not None:
-            code.putln('#define %s %s->%s' % (
-                Naming.preimport_cname,
-                Naming.modulestateglobal_cname,
-                Naming.preimport_cname))
         code.putln('#ifdef __Pyx_CyFunction_USED')
-        code.putln('#define %s %s->%s' % (
+        code.putln('#define %s __Pyx_Globals()->%s' % (
             Naming.cyfunction_type_cname,
-            Naming.modulestateglobal_cname,
             Naming.cyfunction_type_cname))
         code.putln('#endif')
         code.putln('#ifdef __Pyx_FusedFunction_USED')
-        code.putln('#define %s %s->%s' %
+        code.putln('#define %s __Pyx_Globals()->%s' %
             (Naming.fusedfunction_type_cname,
-            Naming.modulestateglobal_cname,
             Naming.fusedfunction_type_cname))
         code.putln('#endif')
 
-    def generate_module_state_clear(self, env, code):
+    def generate_module_state_defines_end(self, env, code):
+        code.putln("#endif /* CYTHON_USE_MODULE_STATE */")
+
+    def generate_module_state_clear_begin(self, env, code):
         code.putln("#if CYTHON_USE_MODULE_STATE")
         code.putln("static int %s_clear(PyObject *m) {" % Naming.module_cname)
-        code.putln("%s *clear_module_state = %s(m);" % (
-            Naming.modulestate_cname,
-            Naming.modulestate_cname))
+        code.putln("%s *clear_module_state = __Pyx_ModuleState(m);" % Naming.modulestate_type_cname)
         code.putln("if (!clear_module_state) return 0;")
         code.putln('Py_CLEAR(clear_module_state->%s);' %
-            Naming.builtins_cname)
+            env.module_dict_cname.field_name)
+        code.putln('Py_CLEAR(clear_module_state->%s);' %
+            Naming.builtins_cname.field_name)
```

Comment:
Yes, I am also contemplating to further raise the abstraction level and maintain a `globals` list in the code environment or so. And then completely generate the `clear` and `visit` calls from that. So that the typical user could would just add a something to the `globals` list once and the rest falls into place automatically.

But again, I'd like to do one step at a time (and have the code testable and working between, so we can hit the master branch rather sooner than later).

Reply:
You mean like what we do in `ModuleNode.generate_module_cleanup_func()`? ;-)

Reply:
That could really be done automatically. When a `Global` declaration is written into the C code of the module (and we could have a dedicated method for that), then it would be kept around and automatically visited/cleared. Sounds simple enough.

Reply:
And we can also post-process the utility code fragments and re-search for `__Pyx_Global(…)`, extract the name, and make sure it gets declared.